### PR TITLE
Allow linking badge templates to regforms

### PR DESCRIPTION
**Pull request by tomasr8**
Something that has been asked many times before is the ability to reference regform fields in the badge designer.
Currently you can use any regform as a source when printing badges and the list of available placeholders is limited to personal data since every regform has them.

The idea is to essentially allow linking templates to a specific regform. Once linked, you would be able reference any field from the linked regform.

Once linked, you have all registration data placeholders available in the badge designer and you can print the badges.

When printing badges/tickets, you can only select the linked regform (if any). The equivalent is true when configuring default ticket template in regforms (TicketsForm) - you can only select unlinked templates, or templates which are linked to the given regform.

This is what the UI for linking looks like currently:
image

For not currently linked templates, there is a 'plus' icon which opens dropdown where you can select to link a regform (same as when printing badges). To unlink, you click on the red 'X'.

### CODE REVIEW

**Comment by ThiefMaster** [comment]
For double-sided:

- Front/back side must be linked to the same regform (this implies both templates are linked to the event itself), or the backside must not be linked to anything
- Linking a template to a regform should not be allowed if it's the backside of another template, unless that template is linked to the same registration form

File: indico/modules/designer/models/templates.py [hunk 1]
```diff
@@ -39,6 +41,12 @@ class DesignerTemplate(db.Model):
         db.String,
         nullable=False
     )
+    registration_form_id = db.Column(
+        db.Integer,
```
Comment (ThiefMaster) [inline indico/modules/designer/models/templates.py:45]: not needed, `db.ForeignKey` looks up the referenced type automatically (which nobody knew early on so in most places we explicitly repeat it)

File: indico/modules/designer/models/templates.py [hunk 2]
```diff
@@ -76,6 +84,15 @@ class DesignerTemplate(db.Model):
         nullable=False,
         default=False
     )
+    regform = db.relationship(
+        'RegistrationForm',
+        lazy=True,
+        foreign_keys=registration_form_id,
```
Comment (ThiefMaster) [inline indico/modules/designer/models/templates.py:90]: is this needed? there's only one FK between those tables

File: indico/modules/designer/placeholders.py [hunk 3]
```diff
@@ -5,10 +5,12 @@
 # modify it under the terms of the MIT License; see the
 # LICENSE file for more details.
 
+import json
 from io import BytesIO
 
 from babel.numbers import format_currency
 from PIL import Image
+from speaklater import _LazyString
```
Comment (ThiefMaster) [inline indico/modules/designer/placeholders.py:13]: `from speaklater import is_lazy_string`

**Comment by bpedersen2** [comment]
Looks promising, if ready I can also try to review it in detail.

**Comment by tomasr8** [comment]
> Looks promising, if ready I can also try to review it in detail.

Hi! Yes, definitely feel free to give it a try :) It should be more or less working at this point, I mostly need to add tests.
I rebased it on the latest master and updated the migration so you just need `indico db upgrade`

File: indico/modules/designer/templates/_list.html [hunk 4]
```diff
@@ -60,9 +68,43 @@
...
+                                            from the template.
+                                      {%- endtrans %}">
+                </button>
+            {% endif %}
+        </td>
         <td>
             <div class="thin toolbar right">
                 <div class="group">
+                    {% if not tpl.registration_form and linkable_forms[tpl] %}
+                        <button class="i-button icon-plus icon-only highlight text-color js-dropdown borderless"
```
Comment (duartegalvao) [inline indico/modules/designer/templates/_list.html:94]: What about using an icon more indicative to what this does... Was thinking either `icon-link` or `icon-clipboard` (which we use for regforms). What do you think?
image

File: indico/modules/designer/templates/_list.html [hunk 5]
```diff
@@ -60,9 +68,43 @@
...
+                <button class="i-button icon-cross icon-only text-color borderless danger"
+                        title="{% trans %}Unlink registration form{% endtrans %}"
+                        data-href="{{ url_for('.unlink_regform', tpl) }}"
+                        data-method="PATCH"
+                        data-update="#template-list"
+                        data-confirm="{% trans title=tpl.registration_form.title -%}
+                                            Are you sure you want to unlink the registration form '{{ title }}'?
+                                            Any placeholders specific to that registration form will be removed
+                                            from the template.
+                                      {%- endtrans %}">
```
Comment (duartegalvao) [inline indico/modules/designer/templates/_list.html:86]: Maybe it would look better with a bit less `padding-left`?
image
(here I'm using `0.1em`)

File: indico/modules/designer/placeholders.py [hunk 6]
```diff
@@ -33,15 +33,17 @@
...
-GROUP_TITLES = {
-    'registrant': _('Registrant Data'),
-    'event': _('Event Data'),
-    'fixed': _('Fixed Data')
+GROUPS = {
+    'event': {'title': _('Event Data'), 'position': 1},
+    'fixed': {'title': _('Fixed Data'), 'position': 2},
+    'registrant': {'title': _('Common Registrant Data'), 'position': 3},
+    # Come up with a better name or merge with registrant data
+    'dynamic': {'title': _('Custom Registrant Data'), 'position': 4},
```
Comment (duartegalvao) [inline indico/modules/designer/placeholders.py:41]: may be a bit wordy but something like `registrant_custom` would at least be more obvious

**Review by duartegalvao — COMMENTED** [review]
seems to work well!

File: indico/modules/designer/models/templates.py [hunk 7]
```diff
@@ -152,6 +168,18 @@ def is_ticket(self):
     def __repr__(self):
         return format_repr(self, 'id', 'event_id', 'category_id', _text=self.title)
 
+    def link_regform(self, regform):
+        """Link this template to a registration form."""
+        self.registration_form = regform
+        db.session.commit()
```
Comment (ThiefMaster) [inline indico/modules/designer/models/templates.py:174]: i'd rather not have commits in such places. we already commit at the end of a request if there's no error, and usually one does not expect some method on a model to commit early

File: indico/modules/designer/__init__.py [hunk 8]
```diff
@@ -80,13 +80,20 @@ class PageLayout(RichIntEnum):
 
 
 @signals.core.get_placeholders.connect_via('designer-fields')
-def _get_notification_placeholders(sender, **kwargs):
+def _get_notification_placeholders(sender, regform=None, **kwargs):
```
Comment (tomasr8) [inline indico/modules/designer/__init__.py:83]: Hmm, why is the method named this way? It has nothing to do with notifications :thinking:

File: indico/modules/designer/templates/_list.html (hunk 4 above)
Comment (tomasr8) [reply indico/modules/designer/templates/_list.html]: I went with the feather icon (`icon-quill`) in the end since that's what we use in most places for regforms

File: indico/modules/designer/placeholders.py [hunk 9]
```diff
@@ -407,3 +408,65 @@ class FixedImagePlaceholder(DesignerPlaceholder):
...
+    def _render_accommodation(self, friendly_data):
+        if friendly_data['is_no_accommodation']:
+            return _('No accommodation')
+
+        arrival = format_date(friendly_data['arrival_date'])  # TODO: include locale
+        departure = format_date(friendly_data['departure_date'])
+        accommodation = friendly_data['choice']
+        return _('''\
+Arrival: {arrival} · Departure {departure} · Accommodation: {accommodation}
+''').format(arrival=arrival, departure=departure, accommodation=accommodation)
```
Comment (ThiefMaster) [inline indico/modules/designer/placeholders.py:458]: I'd rather use a neutral version that does not require i18n, e.g. `{accommodation} ({arrival} - {departure})`.
When someone really puts that field on a badge, they'll probably have to put static text around it anyway.

Because if no event language is set (few events do that) (and even then you'd need to force the event locale which I don't think we do anywhere when printing badges?), using the language of the person generating the badge - in particular if it's a manager - is not ideal. Image a Polish or Chinese user at CERN printing badges for some event happening here... ;)

File: indico/modules/designer/placeholders.py [hunk 10]
```diff
@@ -407,3 +408,65 @@ class FixedImagePlaceholder(DesignerPlaceholder):
     def render(cls, item):
         buf = DesignerImageFile.get(item['image_id']).open()
         return Image.open(buf)
+
+
+class RegistrationFormFieldPlaceholder(DesignerPlaceholder):
```
Comment (ThiefMaster) [inline indico/modules/designer/placeholders.py:413]: It would be nice if this also took the (somewhat new) picture field into account.

Usually this is checked (in various places, also e.g. in the designer to use a square-shaped item) by looking at the class' `is_image`, but of course we cannot do that here because this placeholder can be either an image or nor depending on the field...

File: indico/modules/designer/controllers.py [hunk 11]
```diff
@@ -190,6 +203,7 @@ def clone_template(self, target=None):
         new_template = DesignerTemplate(title=title, type=self.template.type, **target_dict)
 
         data = deepcopy(self.template.data)
+        data = remove_regform_field_placeholders(data)
```
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:206]: I'd rather clone it as linked. It's nice if you can have a backup or want to make just some minor changes!

File: indico/modules/designer/controllers.py [hunk 12]
```diff
@@ -340,6 +356,48 @@ def _process_POST(self):
...
+        return jsonify_data(html=_render_template_list(self.target, event=self.event_or_none))
+
+
+class RHUnLinkDesignerTemplate(RHModifyDesignerTemplateBase):
+    def _process(self):
+        regform = self.template.registration_form
+        if not regform:
+            raise ValueError('This template is not linked to any registration form.')
+
+        self.template.unlink_regform()
```
Comment (ThiefMaster) [inline indico/modules/designer/controllers.py:393]: this is quite destructive, we should have a confirmation prompt in the UI for unlinking. or should we even allow unlinking as long as there are custom placeholders?

Comment (ThiefMaster) [reply indico/modules/designer/controllers.py:393]: ah hm, maybe i unconsciously clicked "yes" :D

anyway, i think preventing is better, i can't see a good reason why you'd add custom fields and then unlink... "least surprise for the user" is generally good. we all know how much users read confirmation prompts (\*cough\* event deletion \*cough\*)

File: indico/migrations/versions/20240314_1430_4e32f4d5ebe4_add_registration_id_to_designer_templates.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+def upgrade():
+    op.add_column('designer_templates', sa.Column('registration_form_id', sa.Integer(), nullable=True), schema='indico')
+    op.create_index(None, 'designer_templates', ['registration_form_id'], schema='indico')
+    op.create_foreign_key(None, 'designer_templates', 'forms', ['registration_form_id'], ['id'],
+                          source_schema='indico', referent_schema='event_registration')
+    op.create_check_constraint('no_regform_if_category', 'designer_templates',
+                               '(category_id IS NULL) OR (registration_form_id IS NULL)', schema='indico')
+
+
+def downgrade():
```
Comment (tomasr8) [inline indico/migrations/versions/20240314_1430_4e32f4d5ebe4_add_registration_id_to_designer_templates.py:28]: Should we delete custom fields from the template data when downgrading? Keeping the data around prevents the template from being modified unless you first delete all custom fields:
image

On the other hand if you ever want to upgrade afterwards you'll lose the data. If we keep the fields you can simply link to the right regform again and it'll work.

Comment (ThiefMaster) [reply indico/migrations/versions/20240314_1430_4e32f4d5ebe4_add_registration_id_to_designer_templates.py:28]: let's keep it, it only breaks saving when you still have the data there, so not a big deal. downgrades are rare anyway, we just want to make sure we do not get a state where indico itself is broken or pages fail on load

File: indico/modules/designer/models/templates.py [hunk 14]
```diff
@@ -152,6 +168,22 @@ def is_ticket(self):
...
         return format_repr(self, 'id', 'event_id', 'category_id', _text=self.title)
 
+    def link_regform(self, regform):
+        """Link this template to a registration form."""
+        self.registration_form = regform
+
+    def unlink_regform(self):
+        """Unlink this template from a registration form."""
+        from indico.modules.designer.util import remove_regform_field_placeholders
+        self.data = remove_regform_field_placeholders(self.data)
```
Comment (ThiefMaster) [inline indico/modules/designer/models/templates.py:178]: Maybe this isn't needed anymore?

File: indico/modules/designer/placeholders.py [hunk 15]
```diff
@@ -422,6 +422,7 @@ class RegistrationFormFieldPlaceholder(DesignerPlaceholder):
     def __init__(self, field):
         self.field = field
         self.name = f'field-{field.id}'
+        self.is_image = self.field.input_type == 'picture'
```
Comment (ThiefMaster) [inline indico/modules/designer/placeholders.py:425]: Nice that this works. I was somehow thinking the signal checks it on the class itself and not on an instance of it. But with instances it's of course easy to handle...

File: indico/modules/designer/__init__.py [hunk 16]
```diff
@@ -80,13 +80,20 @@ class PageLayout(RichIntEnum):
...
         obj = getattr(placeholders, name)
         if isinstance(obj, type) and issubclass(obj, placeholders.DesignerPlaceholder) and hasattr(obj, 'name'):
             yield obj
 
+    if regform:
+        from indico.modules.designer.placeholders import RegistrationFormFieldPlaceholder
+        from indico.modules.events.registration.models.items import RegistrationFormItemType
+        for field in regform.active_fields:
+            if field.type != RegistrationFormItemType.field_pd:      # Personal data fields are already included
+                yield RegistrationFormFieldPlaceholder(field=field)  # in the 'registrant' group
```
Comment (ThiefMaster) [inline indico/modules/designer/__init__.py:95]:
```suggestion
            # Personal data fields are already included in the 'registrant' group
            if field.type != RegistrationFormItemType.field_pd:
                yield RegistrationFormFieldPlaceholder(field=field)
```

not a fan of "block-style" comments ;)

File: indico/modules/designer/templates/_list.html [hunk 17]
```diff
@@ -1,3 +1,11 @@
+{% macro _render_link_regform_button(regform, tpl) %}
+    <a data-href="{{ url_for('designer.link_regform', regform, template_id=tpl.id) }}"
+       data-method="PATCH"
```
Comment (ThiefMaster) [inline indico/modules/designer/templates/_list.html:3]:
```suggestion
       data-method="POST"
```

File: indico/modules/designer/util.py [hunk 18]
```diff
@@ -98,3 +104,55 @@ def get_badge_format(tpl):
...
+    """Return all templates that can be used to print badges/tickets for the given registration form.
+
+    A template can be used if it is either not linked to anything or linked to the current registation form.
+    """
+    return DesignerTemplate.query.with_parent(regform.event).filter(
+        db.or_(
+            DesignerTemplate.registration_form_id.is_(None),
+            DesignerTemplate.registration_form == regform
+        )
+    ).all()
```
Comment (ThiefMaster) [inline indico/modules/designer/util.py:158]: Our usual formatting for multiline queries

```suggestion
    return (
        DesignerTemplate.query
        .with_parent(regform.event)
        .filter(db.or_(DesignerTemplate.registration_form_id.is_(None),
                       DesignerTemplate.registration_form == regform))
        .all()
    )
```

File: indico/modules/designer/blueprint.py [hunk 19]
```diff
@@ -35,6 +36,10 @@ def view_func(**kwargs):
 _bp.add_url_rule('/category/<int:category_id>/manage/designer/<int:template_id>/toggle-default-badge',
                  'toggle_category_default_badge', RHToggleBadgeDefaultOnCategory, methods=('POST',))
 
+_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/link/<int:reg_form_id>', 'link_regform',
+                 RHLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('PATCH',))
+_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/unlink', 'unlink_regform',
+                 RHUnLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('PATCH',))
```
Comment (ThiefMaster) [inline indico/modules/designer/blueprint.py:42]:
```suggestion
_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/link/<int:reg_form_id>', 'link_regform',
                 RHLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('POST',))
_bp.add_url_rule('/event/<int:event_id>/manage/designer/<int:template_id>/unlink', 'unlink_regform',
                 RHUnLinkDesignerTemplate, defaults={'object_type': 'event'}, methods=('POST',))
```

"Non-RESTful" actions usually use POST. It's strange to use PATCH without a payload indicating what to update

File: indico/modules/designer/templates/_list.html [hunk 20]
```diff
@@ -60,9 +68,46 @@
...
+                        {{ tpl.registration_form.title }}
+                    </a>
+                </span>
+                {% set unlinkable = tpl.is_unlinkable %}
+                <button class="i-button icon-cross icon-only text-color borderless danger {{ 'not-unlinkable' if not unlinkable }}"
+                        style="padding-left: 0.2em;"
+                        {% if unlinkable %}
+                            title="{% trans %}Unlink registration form{% endtrans %}"
+                            data-href="{{ url_for('.unlink_regform', tpl) }}"
+                            data-method="PATCH"
```
Comment (ThiefMaster) [inline indico/modules/designer/templates/_list.html:85]:
```suggestion
                            data-method="POST"
```

**Review by ThiefMaster — APPROVED** [review]
